src/platform/pcp.c

```diff
@@ -0,0 +1,574 @@
+            uint8_t Reserved[3];
+            uint16_t InternalPort;
+            uint16_t SuggestedExternalPort;
+            uint8_t SuggestedExternalIpAddress[16];
+            uint16_t RemotePeerPort;
+            uint16_t Reserved2;
+            uint8_t RemotePeerIpAddress[16];
+        } PEER;
+    };
+
+} PCP_REQUEST;
+
+typedef struct PCP_RESPONSE {
+
+    uint8_t Version;
+    uint8_t Opcode : 7;
+    uint8_t Request : 1;
+    uint8_t Reserved1;
+    uint8_t ResultCode;
+    uint32_t Lifetime;
+    uint32_t EpochTime;
+    uint8_t Reserved2[12];
+    union {
+        uint8_t OpcodePayload[0];
+        struct {
+            uint8_t MappingNonce[12];
+            uint8_t Protocol;
+            uint8_t Reserved[3];
+            uint16_t InternalPort;
+            uint16_t AssignedExternalPort;
+            uint8_t AssignedExternalIpAddress[16];
+        } MAP;
+        struct {
+            uint8_t MappingNonce[12];
+            uint8_t Protocol;
+            uint8_t Reserved[3];
+            uint16_t InternalPort;
+            uint16_t AssignedExternalPort;
+            uint8_t AssignedExternalIpAddress[16];
+            uint16_t RemotePeerPort;
+            uint16_t Reserved2;
+            uint8_t RemotePeerIpAddress[16];
+        } PEER;
+    };
+
+} PCP_RESPONSE;
+
+#pragma pack(pop)
+
+#define SIZEOF_THROUGH_FIELD(type, field) \
+    (FIELD_OFFSET(type, field) + sizeof(((type *)0)->field))
+
+const uint16_t PCP_MAP_REQUEST_SIZE = SIZEOF_THROUGH_FIELD(PCP_REQUEST, MAP.SuggestedExternalIpAddress);
+const uint16_t PCP_PEER_REQUEST_SIZE = SIZEOF_THROUGH_FIELD(PCP_REQUEST, PEER.RemotePeerIpAddress);
+
+const uint16_t PCP_MAP_RESPONSE_SIZE = SIZEOF_THROUGH_FIELD(PCP_RESPONSE, MAP.AssignedExternalIpAddress);
+const uint16_t PCP_PEER_RESPONSE_SIZE = SIZEOF_THROUGH_FIELD(PCP_RESPONSE, PEER.RemotePeerIpAddress);
+
+//
+// Main structure for PCP
+//
+typedef struct CXPLAT_PCP {
+
+    void* ClientContext;
+    CXPLAT_PCP_CALLBACK_HANDLER ClientCallback;
+
+    uint32_t GatewayCount;
+
+    _Field_size_(GatewayCount)
+    CXPLAT_SOCKET* GatewaySockets[0];
+
+} CXPLAT_PCP;
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatPcpInitialize(
+    _In_ CXPLAT_DATAPATH* Datapath,
+    _In_ void* Context,
+    _In_ CXPLAT_PCP_CALLBACK_HANDLER Handler,
+    _Out_ CXPLAT_PCP** NewPcpContext
+    )
+{
+    CXPLAT_PCP* PcpContext = NULL;
+    uint32_t PcpContextSize;
+    QUIC_ADDR* GatewayAddresses = NULL;
+    uint32_t GatewayAddressesCount;
+
+    QUIC_STATUS Status =
+        CxPlatDataPathGetGatewayAddresses(
+            Datapath,
+            &GatewayAddresses,
+            &GatewayAddressesCount);
+    if (QUIC_FAILED(Status)) {
+        goto Exit;
+    }
+    CXPLAT_DBG_ASSERT(GatewayAddresses != NULL);
+    CXPLAT_DBG_ASSERT(GatewayAddressesCount != 0);
+
+    PcpContextSize = sizeof(CXPLAT_PCP) + (GatewayAddressesCount * sizeof(CXPLAT_SOCKET*));
+    PcpContext = (CXPLAT_PCP*)CXPLAT_ALLOC_PAGED(PcpContextSize, QUIC_POOL_PCP);
```

Comment:
I'm assuming PAGED is fine here, but then likely couldn't GatewayAddresses also be PAGED. Currently its allocated NONPAGED. Since PAGED memory is less of a system resources issues. (Not in user mode, but still).

Reply:
Actually, this needs to be nonpaged. I'll change it. Good find. (Not that we support kernel yet)